### ISSUE TITLE
Random Forest for Classification

### ISSUE THREAD

**Issue by ConorWaldron** [ConorWaldron/Kaggle#7]

**Comment by ConorWaldron** [comment]
On the training set, the model does very well, with near perfect ROC and PR curves, an F1 score of 0.97 and an accuracy of 0.98. This is expected for a random forest model, the only reason we dont get perfect accuracy is probably because of 2 people who have identical feature vectors but different outcomes so the model cant split them.

![image](https://user-images.githubusercontent.com/48121498/124350626-8c268980-dbed-11eb-9a46-a5954afe4e42.png)
![image](https://user-images.githubusercontent.com/48121498/124350628-8fba1080-dbed-11eb-91a6-65b1ecfc04f9.png)


On the test set the model still does quite well: F1 = 0.75 and Accuracy = 0.81
![image](https://user-images.githubusercontent.com/48121498/124350633-95175b00-dbed-11eb-9ff4-06c570960a51.png)
![image](https://user-images.githubusercontent.com/48121498/124350634-9b0d3c00-dbed-11eb-91f1-7bea00e7e92f.png)


The drop in performance between train and test set is not evidence of over fitting, it is just how random forests work, they get near perfect performance on the train set. They cant actually over fit as rather than extending the decision tree boundaries in weird ways to reach outliers/noise, they wrap them in near infinitely small volume. See read me for more info.

**Comment by ConorWaldron** [comment]
Try changing the default hyper parameters (number of trees etc).

using 1000 trees instead of the default value of 100 increased performance only very marginally
accuracy 0.81 -> 0.82
F1 0.75 -> 0.76
AUC ROC 0.89 -> 0.90
AUC PR 0.86 -> 0.87

**Comment by ConorWaldron** [comment]
Examine difference in performance on training and test set to check for over fitting

**Comment by ConorWaldron** [comment]
Do random forests need data pre processing, check if mean centering the data makes a difference

The internet says random forests do not need data processing (mean scaling or other normalisation methods) as they are decision tree based methods, so it wont affect results at all. 

you just need to one hot encode categorical/binary data

**Comment by ConorWaldron** [comment]
Looking at the feature importances we see the most important variables are you age, sex and how much you paid for your ticket. the class of your ticket matters less than what you paid for it.
![image](https://user-images.githubusercontent.com/48121498/124351121-749cd000-dbf0-11eb-85f0-e92b3b5ba58f.png)

**Comment by ConorWaldron** [comment]
Using AdaBoost model resulted in a significant improvment in the test data preformance

Accuracy 0.81 -> 0.92
F1 0.75 -> 0.88
ROC AUC 0.89 -> 0.93
PR AUC 0.86 -> 0.85

![image](https://user-images.githubusercontent.com/48121498/124357044-58f5f180-dc11-11eb-84e4-99a6dc695da0.png)
![image](https://user-images.githubusercontent.com/48121498/124357046-5bf0e200-dc11-11eb-93bf-e071d83fa55a.png)
![image](https://user-images.githubusercontent.com/48121498/124356958-fd2b6880-dc10-11eb-8e03-be9f505af336.png)
![image](https://user-images.githubusercontent.com/48121498/124356959-00beef80-dc11-11eb-8f56-814f7af541e6.png)
![image](https://user-images.githubusercontent.com/48121498/124356963-03b9e000-dc11-11eb-9c47-92818bb0f1e6.png)

Strangely the AdaBoost performance on the train set is not perfect

Strangely the adaboost performance on the training set is worse than its preformance on the test set. 

The important features for the AdaBoost method change significantly from the random forest, for the Adaboost fare and age are the most important predictors, not gender.

**Comment by ConorWaldron** [comment]
How does feature importance actually work? 

What direction do the features have on the model?
